hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java

```diff
@@ -0,0 +1,664 @@
+      //Lets encode and write
+      //encoder.encode();
+      writeParityCells();
+      // check if block ends?
+      if (currentBlockGroupLen == numDataBlks * blockOutputStreamEntryPool
+          .getStreamEntries().get(blockOutputStreamEntryPool.getCurrIdx())
+          .getLength()) {
+        blockOutputStreamEntryPool.endECBlock(numDataBlks);
+        currentBlockGroupLen = 0;
+      }
+    }
+  }
+
+  void writeParityCells() throws IOException {
+    final ByteBuffer[] buffers = ecChunkBufferCache.getBuffers();
+    //encode the data cells
+    for (int i = 0; i < numDataBlks; i++) {
+      buffers[i].flip();
+    }
+    encode(encoder, numDataBlks, buffers);
+    for (int i = numDataBlks; i < numDataBlks + 2; i++) {
+      handleWrite(buffers[i].array(), 0, cellSize, true, true);
+    }
+
+    ecChunkBufferCache.flipAllDataBuffers();
+    ecChunkBufferCache.clear();
+  }
+
+  private static void encode(RawErasureEncoder encoder, int numData,
+      ByteBuffer[] buffers) throws IOException {
+    final ByteBuffer[] dataBuffers = new ByteBuffer[numData];
+    final ByteBuffer[] parityBuffers = new ByteBuffer[buffers.length - numData];
+    System.arraycopy(buffers, 0, dataBuffers, 0, dataBuffers.length);
+    System.arraycopy(buffers, numData, parityBuffers, 0, parityBuffers.length);
+
+    encoder.encode(dataBuffers, parityBuffers);
+  }
+
+  private void handleWrite(byte[] b, int off, long len, boolean isFullCell,
+      boolean isParity) throws IOException {
+    ecChunkBufferCache
+        .addTo(blockOutputStreamEntryPool.getCurrIdx(), b, off, (int) len);
+    while (len > 0) {
+      try {
+
+        BlockOutputStreamEntry current =
+            blockOutputStreamEntryPool.allocateBlockIfNeeded();
+        // length(len) will be in int range if the call is happening through
+        // write API of blockOutputStream. Length can be in long range if it
+        // comes via Exception path.
+        int expectedWriteLen =
+            Math.min((int) len, (int) current.getRemaining());
+        long currentPos = current.getWrittenDataLength();
+        // writeLen will be updated based on whether the write was succeeded
+        // or if it sees an exception, how much the actual write was
+        // acknowledged.
+        int writtenLength =
+            writeToOutputStream(current, len, b, expectedWriteLen, off,
+                currentPos, isParity);
+        currentBlockGroupLen += isParity ? 0 : writtenLength;
+        if (current.getRemaining() <= 0) {
+          // since the current block is already written close the stream.
+          closeCurrentStream(StreamAction.CLOSE);
+        }
+
+        len -= writtenLength;
+        off += writtenLength;
+
+      } catch (Exception e) {
+        markStreamClosed();
+        throw new IOException(e);
+      }
+
+      if (isFullCell) {
+        handleFlushOrClose(StreamAction.FLUSH);
+        blockOutputStreamEntryPool.updateToNextStream(numDataBlks + 2);
```

Comment:
This "2" should be "numParityBlks" I think.

Reply:
Thanks. I corrected it. It was left by mistake.